File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

/*알고리즘
1. 방법의 수를 세는 count 선언
2. DFS 실행하기(numbers, target, index, sum)
3. count 리턴하기

4. 별도 메서드 : DFS
- 현재의 인덱스가 numbers 배열의 끝까지 도달한 경우
    - sum이 타겟 넘버와 일치한다면 -> count++;
*/


class Solution {
    
    static int count;
    
    public int solution(int[] numbers, int target) {
        
        count = 0;
        
        dfs(numbers, target, 0, 0);
        return count;
    }
    
    public static void dfs(int[] numbers, int target, int index, int sum) {
        if (index == numbers.length) {
            if (sum == target) {
                count++;
            }
            return;
        }
        dfs(numbers, target, index+1, sum + numbers[index]);
        dfs(numbers, target, index+1, sum - numbers[index]);
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
# 알고리즘
# 1. 각 숫자 앞에 연산자를 붙이는 경우의 수 : 2가지(+, -) -> 전체 경우의 수는 2^n
# 2. dfs(index, total) : index번째 수까지 고려했을 때, 누적합이 total인 경우 탐색
# 3. 마지막까지 탐색했을 때 누적합이 target이면 카운팅

def solution(numbers, target):
    answer = [0]
    
    def dfs(index, total):
        if index == len(numbers):
            if total == target:
                answer[0] += 1
            return
        # 2가지 경우의 수
        dfs(index+1, total + numbers[index])
        dfs(index+1, total - numbers[index])

    dfs(0, 0)
    
    return answer[0]